applications/MeshingApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,593 @@
+        N_k(0,3) = 2.0 * normal[1] * normal[2];
+        N_k(0,4) = 2.0 * normal[2] * normal[0];
+        N_k(0,5) = 2.0 * normal[0] * normal[1];
+
+        // Set tangential vectors
+        array_1d<double, 3> t1(3, 0.0);
+        array_1d<double, 3> t2(3, 0.0);
+
+        const double tolerance = std::numeric_limits<double>::epsilon();
+        if(std::abs(normal[0]) > tolerance || std::abs(normal[1]) > tolerance) {
+            const double norm = std::sqrt((t1[0]*t1[0]+t1[1]*t1[1]));
+            t1[0] = normal[1]/norm;
+            t1[1] = -normal[0]/norm;
+
+            t2[0] = -normal[0]*normal[2]/norm;
+            t2[1] = -normal[1]*normal[2]/norm;
+            t2[2] = normal[0]*normal[0]+normal[1]*normal[1]/norm;
+        } else{
+            t1[0] = 1.0;
+            t2[1] = 1.0;
+        }
+
+        T_k1(0,0) = normal[0]*t1[0];
+        T_k1(0,1) = normal[1]*t1[1];
+        T_k1(0,2) = normal[2]*t1[2];
+        T_k1(0,3) = normal[1]*t1[2]+normal[2]*t1[1];
+        T_k1(0,4) = normal[2]*t1[0]+normal[0]*t1[2];
+        T_k1(0,5) = normal[0]*t1[1]+normal[1]*t1[0];
+        
+        T_k2(0,0) = normal[0]*t2[0];
+        T_k2(0,1) = normal[1]*t2[1];
+        T_k2(0,2) = normal[2]*t2[2];
+        T_k2(0,3) = normal[1]*t2[2]+normal[2]*t2[1];
+        T_k2(0,4) = normal[2]*t2[0]+normal[0]*t2[2];
+        T_k2(0,5) = normal[0]*t2[1]+normal[1]*t2[0];
+    }
+    
+    noalias(A1) = prod(trans(p_k),trans(N_k));
+    noalias(A2) = prod(N_k,p_k);
+    noalias(A) += mPenaltyNormal * prod(A1, A2);
+
+    noalias(A1) = prod(trans(p_k),trans(T_k1));
+    noalias(A2) = prod(T_k1,p_k);
+    noalias(A) += mPenaltyTangent*prod(A1, A2);
+
+    noalias(b) += mPenaltyNormal*prod(trans(p_k),trans(N_k)) * itNode->GetValue(CONTACT_PRESSURE);
+
+    //PART 2: Contributions from contact nodes: regard all nodes from the patch which are in contact
+    // Patch center node:
+    if (itPatchNode->Has(CONTACT_PRESSURE)){
+        const array_1d<double, 3>& normal_patch_node = itPatchNode->GetValue(NORMAL);
+        p_k(0,1)=0.0;
+        p_k(0,2)=0.0;
+        p_k(1,4)=0.0;
+        p_k(1,5)=0.0;
+        p_k(2,7)=0.0;
+        p_k(2,8)=0.0;
+        N_k(0,0) = normal_patch_node[0]*normal_patch_node[0];
+        N_k(0,1) = normal_patch_node[1]*normal_patch_node[1];
+        N_k(0,2) = 2*normal_patch_node[0]*normal_patch_node[1];
+        T_k1(0,0) = normal_patch_node[0]*normal_patch_node[1];
+        T_k1(0,1) = -normal_patch_node[0]*normal_patch_node[1];
+        T_k1(0,2) = normal_patch_node[1]*normal_patch_node[1]-normal_patch_node[0]*normal_patch_node[0];
+
+        noalias(A1) = prod(trans(p_k),trans(N_k));
+        noalias(A2) = prod(N_k,p_k);
+        noalias(A) += mPenaltyNormal*prod(A1, A2);
+
+        noalias(A1) = prod(trans(p_k),trans(T_k1));
+        noalias(A2) = prod(T_k1,p_k);
+        noalias(A) += mPenaltyTangent*prod(A1, A2);
+        //noalias(A) += mPenaltyNormal*prod(prod(trans(p_k),trans(N_k)),prod(N_k,p_k));
```

Comment:
?

Reply:
This is the computation of a local matrix

Reply:
same as above, please add a comment